File: decode.cpp
#include <opencv2/opencv.hpp>
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <filesystem>
#include <cstdlib>
#include <cstring>
#include <stdexcept>
#include <sys/stat.h>
#include <unistd.h>

// Alias filesystem namespace for portability
namespace fs = std::__fs::filesystem; // or use std::filesystem if available

// Clear or create a directory
void clear_or_create_dir(const std::string& directory) {
    if (fs::exists(directory)) {
        fs::remove_all(directory);
    }
    fs::create_directories(directory);
}

// Extract frames from video using FFmpeg
void extract_frames_from_video(const std::string& video_file, const std::string& output_dir) {
    clear_or_create_dir(output_dir);
    
    // FFmpeg command to extract frames
    std::string ffmpeg_command = "ffmpeg -i " + video_file + " -vsync 0 " + output_dir + "/frame_%04d.png";
    
    if (system(ffmpeg_command.c_str()) == 0) {
        std::cout << "Frames extracted to: " << output_dir << std::endl;
    } else {
        std::cerr << "Error extracting frames." << std::endl;
    }
}

// Infer file details from the video filename (format: filename-fileextension-filesize.mkv)
std::pair<std::string, int> infer_file_details_from_filename(const std::string& video_filename) {
    std::string base_name = fs::path(video_filename).filename().string();
    size_t last_dash_pos = base_name.rfind('-');
    size_t second_last_dash_pos = base_name.rfind('-', last_dash_pos - 1);
    size_t dot_pos = base_name.rfind('.');

    // Ensure the file ends with ".mkv", has two dashes and the last part is numeric
    if (dot_pos == std::string::npos || base_name.substr(dot_pos + 1) != "mkv" ||
        last_dash_pos == std::string::npos || second_last_dash_pos == std::string::npos) {
        throw std::runtime_error("Filename " + base_name + " is not in the expected format 'filename-fileextension-filesize.mkv'");
    }

    std::string filename = base_name.substr(0, second_last_dash_pos);
    std::string file_extension = base_name.substr(second_last_dash_pos + 1, last_dash_pos - second_last_dash_pos - 1);

    // Attempt to parse the size (last part before ".mkv") as an integer
    int file_size;
    try {
        file_size = std::stoi(base_name.substr(last_dash_pos + 1, dot_pos - last_dash_pos - 1));
    } catch (const std::invalid_argument&) {
        throw std::runtime_error("Unable to parse file size from the filename: " + base_name);
    }

    std::string original_filename = "recovered_" + filename + "." + file_extension;
    return {original_filename, file_size};
}

// Convert extracted frames back into the original file
void images_to_file(const std::string& input_dir, const std::string& output_file, cv::Size frame_size, int original_length, bool save_artifacts) {
    std::vector<unsigned char> decoded_data;

    // Collect all image files
    std::vector<std::string> image_files;
    for (const auto& entry : fs::directory_iterator(input_dir)) {
        if (entry.path().extension() == ".png") {
            image_files.push_back(entry.path().string());
        }
    }
    std::sort(image_files.begin(), image_files.end());

    for (const std::string& image_file : image_files) {
        cv::Mat frame = cv::imread(image_file, cv::IMREAD_GRAYSCALE);
        if (frame.empty()) {
            std::cerr << "Error reading image: " << image_file << std::endl;
            continue;
        }
        decoded_data.insert(decoded_data.end(), frame.begin<unsigned char>(), frame.end<unsigned char>());
    }

    // Trim to original length
    decoded_data.resize(original_length);

    // Write the data back to a file
    std::ofstream output(output_file, std::ios::binary);
    output.write(reinterpret_cast<const char*>(decoded_data.data()), original_length);
    std::cout << "Recovered file saved as " << output_file << std::endl;

    if (!save_artifacts) {
        fs::remove_all(input_dir);
    }
}

int main(int argc, char* argv[]) {
    if (argc < 2) {
        std::cerr << "Usage: " << argv[0] << " <video_file> [--file-size <size>] [--file-format <format>] [--save-artifacts]" << std::endl;
        return 1;
    }

    std::string video_file = argv[1];
    int file_size = 0;
    std::string file_format;
    bool save_artifacts = false;

    // Parse optional arguments
    for (int i = 2; i < argc; ++i) {
        if (std::strcmp(argv[i], "--file-size") == 0 && i + 1 < argc) {
            file_size = std::stoi(argv[++i]);
        } else if (std::strcmp(argv[i], "--file-format") == 0 && i + 1 < argc) {
            file_format = argv[++i];
        } else if (std::strcmp(argv[i], "--save-artifacts") == 0) {
            save_artifacts = true;
        }
    }

    std::string original_filename;
    if (file_size > 0 && !file_format.empty()) {
        original_filename = "recovered_file." + file_format;
        std::cout << "Using provided size and format: " << original_filename << ", Size: " << file_size << " bytes" << std::endl;
    } else {
        try {
            auto [inferred_filename, inferred_size] = infer_file_details_from_filename(video_file);
            original_filename = inferred_filename;
            file_size = inferred_size;
            std::cout << "Inferred original file: " << original_filename << ", Size: " << file_size << " bytes" << std::endl;
        } catch (const std::runtime_error& e) {
            std::cerr << e.what() << std::endl;
            std::cerr << "Error: Please provide --file-size and --file-format or rename the input file to follow the format 'filename-fileextension-filesize.mkv'." << std::endl;
            return 1;
        }
    }

    // Step 2: Extract frames from video
    std::string output_images_dir = "decoded_images";
    extract_frames_from_video(video_file, output_images_dir);

    // Step 3: Reconstruct the original file from the frames
    images_to_file(output_images_dir, original_filename, cv::Size(640, 480), file_size, save_artifacts);

    return 0;
}

File: decode.py
import cv2
import numpy as np
import os
import shutil
import subprocess
import argparse

def clear_or_create_dir(directory):
    """Empty the directory if it exists, otherwise create it."""
    if os.path.exists(directory):
        # Remove all contents in the directory
        shutil.rmtree(directory)
    # Recreate the directory
    os.makedirs(directory)

def extract_frames_from_video(video_file, output_dir):
    """Run FFmpeg to extract frames from video."""
    # Clear or create the output directory
    clear_or_create_dir(output_dir)

    # FFmpeg command to extract frames from video
    ffmpeg_command = [
        'ffmpeg',
        '-i', video_file,              # Input video file
        '-vsync', '0',                 # Disable frame rate synchronization
        f'{output_dir}/frame_%04d.png' # Output file pattern
    ]

    # Run the FFmpeg command using subprocess
    try:
        subprocess.run(ffmpeg_command, check=True)
        print(f"Frames extracted to: {output_dir}")
    except subprocess.CalledProcessError as e:
        print(f"Error: {e}")

def infer_file_details_from_filename(video_filename):
    """
    Infer the original file's name, format, and size from the video filename.
    The filename format should be: filename-extension-size.mkv
    """
    # Extract the base filename without the extension
    base_name = os.path.basename(video_filename)
    name_parts = base_name.rsplit('-', 2)  # Split based on the last two dashes

    if len(name_parts) != 3 or not name_parts[2].endswith('.mkv'):
        raise ValueError(f"Filename {base_name} is not in the expected format: 'filename-extension-size.mkv'")

    # Extract filename, extension, and size
    filename = name_parts[0]
    file_extension = name_parts[1]
    file_size = int(name_parts[2].replace('.mkv', ''))  # Remove the '.mkv' part

    # Reconstruct the original file name with the extension
    original_filename = f"recovered_{filename}.{file_extension}"
    
    return original_filename, file_size

def images_to_file(input_dir, output_file, frame_size=(640, 480), original_length=None):
    """Convert the extracted frames back into the original file."""
    decoded_data = []

    # List all image files in the directory
    image_files = sorted([f for f in os.listdir(input_dir) if f.endswith('.png')])

    for image_file in image_files:
        # Read each frame as a grayscale image
        frame = cv2.imread(os.path.join(input_dir, image_file), cv2.IMREAD_GRAYSCALE)
        decoded_data.append(frame.flatten())

    # Concatenate the data and trim to the original length
    decoded_data = np.concatenate(decoded_data, axis=0)
    byte_data = decoded_data[:original_length].astype(np.uint8).tobytes()

    # Write the byte data back to the file
    with open(output_file, 'wb') as f:
        f.write(byte_data)

    print(f"Recovered file saved as {output_file}")

if __name__ == "__main__":
    # Parse command-line arguments
    parser = argparse.ArgumentParser(description='Recover original file from a video.')
    parser.add_argument('video_file', type=str, help='The video file to decode.')
    parser.add_argument('--file-size', type=int, help='The original file size in bytes.')
    parser.add_argument('--file-format', type=str, help='The original file format (e.g., zip, pdf).')

    args = parser.parse_args()

    # Step 1: If --file-size and --file-format are provided, use them; otherwise, try to infer
    if args.file_size and args.file_format:
        original_filename = f"recovered_file.{args.file_format}"
        original_size = args.file_size
        print(f"Using provided size and format: {original_filename}, Size: {original_size} bytes")
    else:
        try:
            original_filename, original_size = infer_file_details_from_filename(args.video_file)
            print(f"Inferred original file: {original_filename}, Size: {original_size} bytes")
        except ValueError as e:
            # If we can't infer the details and no size or format is provided, output an error
            print(e)
            print("Error: Please provide --file-size and --file-format or rename the input file to follow the format 'filename-extension-size.mkv'.")
            exit(1)

    # Step 2: Extract frames from the video
    output_images_dir = 'decoded_images'
    extract_frames_from_video(args.video_file, output_images_dir)

    # Step 3: Reconstruct the original file from the frames
    images_to_file(output_images_dir, original_filename, frame_size=(640, 480), original_length=original_size)
